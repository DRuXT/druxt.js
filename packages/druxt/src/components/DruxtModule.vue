<script>
import { pascalCase, splitByCase } from 'scule'
import Vue from 'vue'

import DruxtWrapper from './DruxtWrapper.vue'

/**
 * The DruxtModule base Vue.js component.
 *
 * Extend this component to build a Druxt module.
 *
 * @example @lang js
 * import DruxtModule from 'druxt/dist/components/DruxtModule.vue'
 * export default {
 *   name: 'DruxtTestModule',
 *   extends: DruxtModule,
 *   druxt: {
 *     componentOptions: () => ([['wrapper']]),
 *     propsData: (ctx) => ({
 *       bar: ctx.bar,
 *       foo: ctx.foo,
 *     }),
 *   }
 * }
 */
export default {
  components: { DruxtWrapper },

  /** */
  props: {
    /**
     * The module value.
     *
     * @type {(Array|Boolean|Date|Number|Object|String)}
     * @model
     */
    value: {
      type: [Array, Boolean, Date, Number, Object, String],
      default: null,
    },

    wrapper: {
      type: Object,
      default: () => ({
        component: 'div',
        propsData: {},
      })
    },
  },

  /**
   * Loads the Druxt module data and applies a wrapper component as required.
   *
   * **Important:** If your component has an existing `fetch` method, you must manually invoke
   * the `DruxtModule.fetch()` hook.
   *
   * @example @lang js <caption>Manually invoking DruxtModule.fetch().</caption>
   * import DruxtModule from 'druxt/dist/components/DruxtModule.vue'
   * export default {
   *   name: 'DruxtTestModule',
   *   extends: DruxtModule,
   *   async fetch() {
   *     await DruxtModule.fetch.call(this)
   *   }
   *   druxt: {
   *     componentOptions: () => ([['wrapper']]),
   *     propsData: (ctx) => ({
   *       bar: ctx.bar,
   *       foo: ctx.foo,
   *     }),
   *   }
   * }
   */
  async fetch() {
    if (!(this.$options || {}).druxt) {
      return false
    }

    // Build wrapper component object.
    const options = this.getModuleComponents()
    let component = {
      is: (((options.filter(o => o.global) || [])[0] || {}).name || 'DruxtWrapper'),
      options: options.map(o => o.name) || [],
    }

    // Get scoped slots.
    component.slots = Object.keys(this.getScopedSlots())

    // Get wrapper data.
    const wrapperData = await this.getWrapperData(component.is)
    component.settings = wrapperData.druxt || {}

    // Build wrapper component propsData.
    component = { ...component, ...this.getModulePropsData(wrapperData.props) }

    // Set component data.
    this.component = component
  },

  /**
   * @property {ComponentData} component - The wrapper component and propsData to be rendered.
   * @property {object} model - The model object.
   */
  data: ({ value }) => ({
    component: {
      $attrs: {},
      is: 'DruxtWrapper',
      options: [],
      props: {},
      propsData: {},
      settings: {},
      slots: [],
    },
    model: value,
  }),

  /** */
  methods: {
    /**
     * Get list of module wrapper components.
     *
     * @returns {Components}
     */
    getModuleComponents() {
      if (!(this.$options.druxt || {}).componentOptions) {
        return []
      }

      const options = this.$options.druxt.componentOptions.call(this, this)
      if (!options || !options.length) {
        return []
      }

      // Build list of available components.
      let components = []
      for (const set of options.filter(set => Array.isArray(set))) {
        const variants = []
        components = [...components, ...set.map(item => {
          // Build array of name parts.
          const parts = variants.length ? [...variants[0].parts] : []
          parts.push(pascalCase(splitByCase(item)))

          // Convert parts into a pascalCase component name.
          const name = pascalCase([this.$options.name, ...parts])

          // Check if component is globally registered.
          const global = !!this.$options.components[name]

          // Store set variant data to be used in next set item.
          variants.unshift({ global, name, parts })

          return { global, name, parts }
        })]
      }

      // Filter unique components.
      const unique = components.filter(((s) => (o) => !s.has(o.name) && s.add(o.name))(new Set))

      // Sort items by parts length.
      const sorted = unique.sort((a, b) => b.parts.length - a.parts.length)

      return sorted
    },

    /**
     * Get module propsData via modules `druxt.propsData()` callback.
     *
     * @example @lang js
     * {
     *   bar: 'foo',
     *   foo: 'bar',
     * }
     *
     * @return {object}
     */
    getModulePropsData(wrapperProps = {}) {
      if (!(this.$options.druxt || {}).propsData) {
        return {}
      }

      const propsData = this.$options.druxt.propsData.call(this, this)

      // Props.
      const props = {}
      const propsKeys = Object.keys(wrapperProps).filter(i => Object.keys(propsData).includes(i))
      for (const key of propsKeys) {
        props[key] = propsData[key]
      }

      // $attrs.
      const $attrs = { ...this.$attrs }
      delete $attrs['data-fetch-key']
      const $attrsKeys = Object.keys(propsData).filter(i => !Object.keys(wrapperProps).includes(i))
      for (const key of $attrsKeys) {
        $attrs[key] = propsData[key]
      }

      return { $attrs, props, propsData }
    },

    /**
     * Gets a Druxt modules scoped slots, and if there's no default slots,
     * provides a develop mode debug default or passes through to a 
     * default template.
     *
     * @return {object}
     */
    getScopedSlots() {
      const h = this.$createElement
      const scopedSlots = typeof (this.$options.druxt || {}).slots === 'function'
        ? this.$options.druxt.slots.call(this, h)
        : {}

      // Pass through default scoped slot if provided.
      if (typeof this.$scopedSlots.default === 'function') {
        scopedSlots.default = (attrs) => this.$scopedSlots.default({
          ...((this.$options.druxt || {}).propsData || (() => {}))(this),
          ...attrs
        })
      }

      // Provide debug data if Nuxt is running in dev mode.
      if (!scopedSlots.default && this.$nuxt.context.isDev)  {
        scopedSlots.default = () => h(
          'details',
          {
            style: {
              border: '2px dashed lightgrey',
              margin: '0.5em 0',
              padding: '1em',
            },
          },
          [
            h('summary', [`[${this.$options._componentTag}] Missing default slot.`]),
            h('label', ['Component options:', h('ul', this.component.options.map((s) => h('li', [s])))]),
            h('br'),
            h('label', ['propsData:', h('pre', [h('code', [JSON.stringify(this.component.propsData, null, '\t')])])])
          ]
        )
      }

      return scopedSlots
    },

    /**
     * Get wrapper component data.
     *
     * @param {string} component - The Wrapper component name.
     *
     * @return {WrapperData}
     */
    async getWrapperData(component) {
      let wrapperData = { druxt: {}, props: {} }
      if (!this.$options.components[component]) {
        return wrapperData
      }

      // Get data from resolved component.
      if (this.$options.components[component].options) {
        wrapperData = this.$options.components[component].options
      }

      // Get data from unresolved component.
      else if (typeof this.$options.components[component] === 'function' && this._init) {
        wrapperData = (await this.$options.components[component].call(this)) || {}
      }

      const options = Vue.util.mergeOptions({}, wrapperData)
      return {
        druxt: options.druxt || {},
        props: options.props || {},
      }
    }
  },

  watch: {
    model() {
      if (this.component.props.value !== this.model) {
        this.component.props.value = this.model
       
        // Only emit 'input' if using the default 'DruxtWrapper' component.
        if (this.component.is === 'DruxtWrapper') {
          this.$emit('input', this.model)
        }
      }
    },

    value() {
      if (this.value !== this.model) {
        this.model = this.value
      }
    }
  },

  render(h) {
    const self = this

    const wrapperData = {
      class: this.wrapper.class || undefined,
      style: this.wrapper.style || undefined,
      props: this.wrapper.propsData,
    }

    // Return only wrapper if fetch state is still pending.
    if (this.$fetchState.pending) {
      return h(this.wrapper.component, wrapperData)
    }

    // Return wrapped component.
    const attrs = { ...this.component.$attrs, ...this.$attrs }
    delete attrs['data-fetch-key']
    return h(this.wrapper.component, wrapperData, [
      h(this.component.is, {
        attrs,
        on: {
          input(value) {
            self.model = value
            self.$emit('input', value)
          }
        },
        props: this.component.props,
        ref: 'component',
        scopedSlots: this.getScopedSlots(),
      })
    ])
  }
}

/**
 * @typedef {object[]} Components
 * @property {boolean} global - Component global registration state.
 * @property {string} name - The component name.
 * @property {string[]} parts - The component naming parts.
 *
 * @example @lang js
 * [{
 *   global: true,
 *   pascal: 'DruxtTestModuleWrapper',
 *   parts: ['Wrapper'],
 * }]
 */

/**
 * @typedef {object} ComponentData
 * @property {object} $attrs - propsData not registered by the Wrapper component.
 * @property {string} is=DruxtWrapper - The Wrapper component name.
 * @property {string[]} options - The Wrapper component options.
 * @property {object} props - propsData registered by the Wrapper component.
 * @property {object} propsData - The component propsData object.
 * @property {object} settings - Druxt settings object provided by the Wrapper component.
 *
 * @example @lang js
 * {
 *   $attrs: { bar: 'foo' },
 *   is: 'DruxtTestModuleWrapper',
 *   options: [
 *     'DruxtTestModuleWrapper',
 *   ],
 *   props: { foo: 'bar' },
 *   propsData: {
 *     bar: 'foo',
 *     foo: 'bar',
 *   },
 *   settings: { fooBar: true },
 * }
 */

/**
 * @typedef {object} WrapperData
 * @property {object} druxt - Druxt settings object for use by Druxt module.
 * @property {object} props - Registered props oject.
 *
 * @example @lang js
 * {
 *   druxt: { fooBar: true },
 *   props: {
 *     foo: {
 *       type: String,
 *       default: '',
 *     }
 *   }
 * }
 */
</script>
