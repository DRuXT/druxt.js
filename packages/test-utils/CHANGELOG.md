# druxt-test-utils

## 0.2.15

### Patch Changes

- Updated dependencies [41cab3a0]
  - druxt@0.22.0

## 0.2.14

### Patch Changes

- Updated dependencies [e46a3290]
  - druxt@0.21.0

## 0.2.13

### Patch Changes

- Updated dependencies [49b67872]
  - druxt@0.20.0

## 0.2.12

### Patch Changes

- Updated dependencies [be21952]
  - druxt@0.19.0

## 0.2.11

### Patch Changes

- Updated dependencies [e6cf1fc]
  - druxt@0.18.3

## 0.2.10

### Patch Changes

- 540afca: Updated drupal-jsonapi-params to 2.0.0
- Updated dependencies [540afca]
  - druxt@0.18.2

## 0.2.9

### Patch Changes

- Updated dependencies [80164a1]
  - druxt@0.18.1

## 0.2.8

### Patch Changes

- Updated dependencies [e3d5238]
  - druxt@0.18.0

## 0.2.7

### Patch Changes

- Updated dependencies [b79701c]
  - druxt@0.17.0

## 0.2.6

### Patch Changes

- Updated dependencies [7b749bd]
  - druxt@0.16.0

## 0.2.5

### Patch Changes

- Updated dependencies [2ae1d6d]
  - druxt@0.15.0

## 0.2.4

### Patch Changes

- Updated dependencies [45bc0b9]
  - druxt@0.14.0

## 0.2.3

### Patch Changes

- Updated dependencies [77ab204]
  - druxt@0.13.0

## 0.2.2

### Patch Changes

- Updated dependencies [4504a2f]
  - druxt@0.12.0

## 0.2.1

### Patch Changes

- Updated dependencies [75ff8a9]
  - druxt@0.11.0

## 0.2.0

### Patch Changes

- Updated dependencies [715e5ef]
  - druxt@0.10.0

## 0.1.0

### Patch Changes

- Updated dependencies [21170fb]
  - druxt@0.9.0
