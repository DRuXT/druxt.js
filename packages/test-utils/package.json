{
  "name": "druxt-test-utils",
  "version": "0.1.0",
  "description": "Test utilities for Druxt development.",
  "keywords": [
    "druxt",
    "jsonapi",
    "test"
  ],
  "exports": {
    ".": {
      "require": "./dist/druxt-test-utils.ssr.js",
      "import": "./dist/druxt-test-utils.esm.js"
    }
  },
  "main": "dist/druxt-test-utils.ssr.js",
  "module": "dist/druxt-test-utils.esm.js",
  "dependencies": {
    "drupal-jsonapi-params": "^1.2.2",
    "jest-mock-axios": "^4.4.0"
  },
  "peerDependencies": {
    "druxt": "*"
  }
}
