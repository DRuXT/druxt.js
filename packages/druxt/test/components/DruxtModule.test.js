import 'regenerator-runtime/runtime'
import { createLocalVue, mount } from '@vue/test-utils'

import DruxtModule from '../../src/components/DruxtModule.vue'

// Setup local vue instance.
const localVue = createLocalVue()

let mocks

describe('DruxtModule component', () => {
  beforeEach(() => {
    mocks = {
      $createElement: jest.fn(),
      $fetchState: { pending: false },
      $nuxt: {
        context: {
          isDev: false,
        },
      },
    }
  })

  test('defaults', async () => {
    mocks.$fetchState.pending = true
    const wrapper = mount(DruxtModule, { localVue, mocks })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Data.
    expect(wrapper.vm.component).toStrictEqual({
      $attrs: {},
      is: 'DruxtWrapper',
      options: [],
      props: {},
      propsData: {},
      settings: {},
      slots: [],
    })

    // Methods.
    const wrapperData = await wrapper.vm.getWrapperData(wrapper.vm.component.is)
    expect(wrapperData).toStrictEqual({ druxt: {}, props: {} })

    expect(wrapper.vm.getModuleComponents()).toStrictEqual([])
    expect(wrapper.vm.getModulePropsData(wrapperData.props)).toStrictEqual({})

    const scopedSlots = wrapper.vm.getScopedSlots()
    expect(Object.keys(scopedSlots)).toStrictEqual([])

    const mock = {
      _init: jest.fn(),
      $options: { components: { test: jest.fn(() => ({ druxt: {}, props: {} })) }
    }}
    expect(await DruxtModule.methods.getWrapperData.call(mock, 'test')).toStrictEqual({ druxt: {}, props: {} })

    // HTML snapshot.
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('defaults - v-model', async () => {
    const Component = {
      template: "<DruxtModule v-model='model' ref='module' />",
      components: { DruxtModule },
      data: () => ({ model: null })
    }
    const wrapper = mount(Component, { localVue, mocks, stubs: ['DruxtWrapper'] })

    // Default state.
    expect(wrapper.vm.model).toStrictEqual(null)
    expect(wrapper.vm.$refs.module.component.props.value).toStrictEqual(undefined)
    expect(wrapper.vm.$refs.module.model).toStrictEqual(null)
    expect(wrapper.vm.$refs.module.value).toStrictEqual(null)

    // Change model value.
    await wrapper.setData({ model: { test: true }})

    expect(wrapper.vm.model).toStrictEqual({ test: true })
    expect(wrapper.vm.$refs.module.component.props.value).toStrictEqual({ test: true })
    expect(wrapper.vm.$refs.module.model).toStrictEqual({ test: true })
    expect(wrapper.vm.$refs.module.value).toStrictEqual({ test: true })
  })

  test('custom module - no wrapper', async () => {
    const CustomModule = {
      name: 'CustomModule',
      extends: DruxtModule,
      druxt: {
        componentOptions: () => {},
        propsData: () => ({ foo: 'bar' }),
        slots: (h) => ({ default: () => h('div', ['test'] )}),
      }
    }

    const wrapper = mount(CustomModule, { localVue, mocks, stubs: ['DruxtWrapper'] })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Data.
    expect(wrapper.vm.component).toStrictEqual({
      $attrs: { foo: 'bar' },
      is: 'DruxtWrapper',
      options: [],
      props: {},
      propsData: { foo: 'bar' },
      settings: {},
      slots: ['default'],
    })

    // Methods.
    const wrapperData = await wrapper.vm.getWrapperData(wrapper.vm.component.is)
    expect(wrapperData).toStrictEqual({ druxt: {}, props: {} })

    expect(wrapper.vm.getModuleComponents(wrapperData.props)).toStrictEqual([])
    expect(wrapper.vm.getModulePropsData()).toStrictEqual({
      $attrs: { foo: 'bar' },
      props: {},
      propsData: { foo: 'bar' },
    })
    expect(Object.keys(wrapper.vm.getScopedSlots())).toStrictEqual(['default'])

    // HTML snapshot.
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('custom module - wrapper', async () => {
    localVue.component('CustomModuleWrapper', {
      druxt: { foo: 'bar' },
      props: ['foo'],
      render: () => {}
    })

    const CustomModule = {
      name: 'CustomModule',
      extends: DruxtModule,
      druxt: {
        componentOptions: () => ([['wrapper']]),
        propsData: () => ({ foo: 'bar' }),
        slots: (h) => ({ default: () => h('div', ['test'] )}),
      }
    }

    const wrapper = mount(CustomModule, { localVue, mocks, stubs: ['DruxtWrapper'] })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Data.
    expect(wrapper.vm.component).toStrictEqual({
      $attrs: {},
      is: 'CustomModuleWrapper',
      options: ['CustomModuleWrapper'],
      props: { foo: 'bar' },
      propsData: { foo: 'bar' },
      settings: { foo: 'bar' },
      slots: ['default'],
    })

    // Methods.
    const wrapperData = await wrapper.vm.getWrapperData(wrapper.vm.component.is)
    expect(wrapperData.druxt).toStrictEqual({ foo: 'bar' })
    expect(Object.keys(wrapperData.props)).toStrictEqual(['foo'])

    expect(wrapper.vm.getModuleComponents()).toStrictEqual([{
      global: true,
      name: 'CustomModuleWrapper',
      parts: ['Wrapper']
    }])
    expect(wrapper.vm.getModulePropsData(wrapperData.props)).toStrictEqual({
      $attrs: {},
      props: { foo: 'bar' },
      propsData: { foo: 'bar' },
    })
    expect(Object.keys(wrapper.vm.getScopedSlots())).toStrictEqual(['default'])

    // v-model
    expect(wrapper.vm.model).toBe(null)
    await wrapper.setData({ model: { test: 'true' } })
    expect(wrapper.vm.model).toStrictEqual({ test: 'true' })

    wrapper.vm.$refs.component.$emit('input', { test: false })
    expect(wrapper.vm.model).toStrictEqual({ test: false })

    // HTML snapshot.
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('dev mode slot', async () => {
    mocks.$nuxt.context.isDev = true
    const wrapper = mount(DruxtModule, { localVue, mocks })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    // Default slot.
    await wrapper.setData({ component: {
      ...wrapper.vm.component,
      options: ['Test'],
    } })
    const slot = wrapper.vm.getScopedSlots().default.call(wrapper.vm)
    expect(slot.tag).toBe('details')
  })

  test('custom default slot', async () => {
    const scopedSlots = { default: jest.fn() }
    const wrapper = mount(DruxtModule, { localVue, mocks, scopedSlots })
    await wrapper.vm.$options.fetch.call(wrapper.vm)

    wrapper.vm.getScopedSlots().default.call(wrapper.vm)
    expect(scopedSlots.default).toHaveBeenCalled()
  })
})
