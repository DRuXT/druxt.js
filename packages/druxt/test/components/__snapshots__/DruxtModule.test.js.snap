// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DruxtModule component custom module - no wrapper 1`] = `
"<div>
  <druxtwrapper-stub foo=\\"bar\\"></druxtwrapper-stub>
</div>"
`;

exports[`DruxtModule component custom module - wrapper 1`] = `
"<div>
  <!---->
</div>"
`;

exports[`DruxtModule component defaults 1`] = `"<div></div>"`;
